harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+                            "agentId": self.agent_id,
+                            "gpuUuid": gpu_uuid,
+                            "metricType": SYSTEM_METRIC_TYPE_ENUM
+                        }
+                    }
+                    for measurement in self.batch[metric_type][gpu_uuid]:
+                        single_metric_batch["values"].append(measurement.measurement)
+                        single_metric_batch["batches"].append(measurement.batch_index)
+                        single_metric_batch["timestamp"].append(self.convert_to_timestamp_str(measurement.timestamp))
+                    post_formatted.append(single_metric_batch)
+        return post_formatted
+
+
+class SysMetricCollectorThread(threading.Thread):
+    """
+    Background thread for collecting profiler metrics at a high granularity and shipping them to the master
+
+    - SimpleCpuUtilization = Measured in percent
+    - FreeMemory = Measured in Gigabytes
+    - NetworkSentThroughput = Measured in Gigabit/s
+    - NetworkRecvThroughput = Measured in Gigabit/s
+    - DiskIops
+    - DiskReadThroughput = Measured in bytes/second
+    - DiskWriteThroughput = Measured in bytes/second
+    - GpuUtilization = Measured in percent
+    """
+
+    ACTIVE_POLL_INTERVAL = 1  # Check if metric collection has been turned on/off every 1 second
+    FLUSH_INTERVAL = 10  # How often to make API calls
+    MEASUREMENT_INTERVAL = 0.1
+
+    def __init__(self, trial_id, agent_id, send_queue):
+
+        self.is_active = False
+        self.quitting = False
+
+        self.current_batch_idx = 0
+        self.send_queue = send_queue
+        self.current_batch = SysMetricBatcher(trial_id, agent_id)
+        self.current_batch.clear()
+
+        super().__init__()
+
+    def activate(self):
+        self.is_active = True
+
+    def kill(self):
+        self.quitting = True
+
+    def update_batch_idx(self, new_batch_idx):
+        self.current_batch_idx = new_batch_idx
+
+    def enqueue_for_async_send(self, metric_batch):
+        # TODO: Handle exception
+        # This method can theoretically raise a FULL error, but SimpleQueues are unbounded so
+        # I don't think it will ever happen (https://docs.python.org/3/library/queue.html#queue.Queue.put)
+        # self.log("Enqueuing metric batch", metric_batch)
+        self.send_queue.put_nowait(metric_batch)
+
+    def run(self) -> None:
+        last_measurement_time = None
+        batch_start_time = None
+        cpu_util_collector = SimpleCpuUtilCollector()
+        gpu_util_collector = GpuUtilCollector()
+        network_throughput_collector = NetThroughputCollector()
+        free_memory_collector = FreeMemoryCollector()
+        disk_collector = DiskReadWriteRateCollector()
+
+        while True:
+            if self.quitting:
+                # We drop any partial batches
+                break
+
+            if not self.is_active:
+                time.sleep(1)
+                continue
```

Comment:
This is going to result in very low granularity of control on startup times.  If you wanted profile, for example, just one batch, this polling strategy won't let you do that at all; you'll likely be done with the batch before the collection thread even responds to `_begin_collection()`.

A more responsive design would be something like:

```python
class ActivateMessage:
    pass


class QuitMessage:
    pass


class ProfilerAgent:
    # ...
    def _begin_collection(self):
        self.sys_metric_collector_thread.msg_queue.put(ActivateMessage())

    def _end_collection(self):
        self.sys_metric_collector_thread.msg_queue.put(QuitMessage())
        # ...
        

class SysMetricCollectorThread(threading.Thread):
    # ...

    def run(self):
        # ...
        msg = self.msg_queue.get()
        if isinstance(msg, QuitMessage):
             # None means "quit"
             return
        
        # avoid drift in period by explicitly calculating timing
        # rather than relying on sleep(period) 
        next_collection = time.time()

        while True:
            now = time.time()
            if now < next_collection:
                # check for quit message while we wait
                sleep_time = next_collection - now
                try:
                    msg = self.msg_queue.get(timeout=sleep_time)
                    if isinstance(msg, QuitMessage):
                        return
                except queue.Empty:
                   pass

            # avoid drift
            next_collection += period

            # ... do collection and batching and such
```

The above pattern will have maximal responsiveness to the control thread.  It uses timeouts on the `Queue.get()` to implement the periodicity, and the only time it's not waiting on the `.get()` is during the time we are actually collecting metrics.  It also has the benefit that it does not have any shared memory at all; you could use multiprocessing Processes just as easily as threading Threads.

It doesn't full solve the `accurately-profile-one-batch` solution though; that would require a synchronization message in the reverse direction.  But it's a good start.

Reply:
Changed